src/uct/tcp/tcp_ep.c

```diff
@@ -69,7 +69,8 @@ static inline int uct_tcp_ep_ctx_buf_need_progress(uct_tcp_ep_ctx_t *ctx)
 static inline ucs_status_t uct_tcp_ep_check_tx_res(uct_tcp_ep_t *ep)
 {
     if (ucs_likely((ep->conn_state == UCT_TCP_EP_CONN_STATE_CONNECTED) &&
-                   uct_tcp_ep_ctx_buf_empty(&ep->tx))) {
+                   uct_tcp_ep_ctx_buf_empty(&ep->tx) &&
+                   (ep->put_cnt != UINT32_MAX))) {
```

Comment:
can you update `ep->ctx` instead so that it will be "not empty" to avoid extra branch?

Reply:
when we increment `ep->put_cnt`, `ep->tx` isn't empty. and when we do progress of TX oeprations we don't really know if it is PUT or some AM operations.
so, we can't do it.

Reply:
we can have some flag which indicates "CAN_SEND" and set it when both are true:
- TX is empty
- put_cnt != UINT32_MAX

Reply:
done